frontend/components/tree/js/tree.service.ts

```diff
@@ -905,29 +931,97 @@ export class TreeService {
 		return !node.children || !node.children || node.children.length === 0;
 	}
 
+	/**
+	 * Handle visibility changes from tree service to viewer service.
+	 * @param clickedIds	Collection of ids to show/hide.
+	 * @param visible	Set ids to visibile.
+	 */
+	public handleVisibility(clickedIds: any, visible: boolean) {
+
+		const objectIds = {};
+
+		for (const id in clickedIds) {
+			if (id) {
+				const account = clickedIds[id].account;
+				const model = clickedIds[id].model || clickedIds[id].project; // TODO: Kill .project from backend
+				const key = account + "@" + model;
+
+				if (!objectIds[key]) {
+					objectIds[key] = [];
+				}
+
+				objectIds[key].push(id);
+			}
+		}
+
+		// Update viewer object visibility
+		for (const key in objectIds) {
+			if (key) {
+				const vals = key.split("@");
+				const account = vals[0];
+				const model = vals[1];
+
+				if (this.ViewerService.viewer) {
+
+					this.ViewerService.switchObjectVisibility(
+						account,
+						model,
+						objectIds[key],
+						visible,
+					);
+				}
+
+			}
+		}
+	}
+
 	/**
 	 * Update the state of clickedHidden and clickedShown, which are used by tree component
 	 * to apply changes to the viewer.
 	 * @param node	Node to toggle visibility. All children will also be toggled.
 	 */
-	public updateModelState(node) {
+	public updateModelVisibility(node) {
 
 		this.ready.promise.then(() => {
-			const childNodes = {};
-			this.traverseNodeAndPushId(node, childNodes, this.treeMap.idToMeshes);
+
+			const childNodes = this.getMeshMapFromNodes(node, this.treeMap.idToMeshes);
```

Comment:
This is the only place where getMeshMapFromNodes is given a single `node` rather than an array of `nodes`. We can pass in `[node]` to save getMeshMapFromNodes having to deal with different inputs.